### PR TITLE
Additional errors

### LINKED ISSUE #92: Add support for v3.0 conditionally required fields
New conditional requirement added [here](https://github.com/MobilityData/gbfs/pull/457/files) for `default_reserve_time` in `vehicle_types.json` that is not supported by the schema:

> REQUIRED if `reservation_price_per_min` or `reservation_price_flat_rate` are defined.

Thanks @isabelle-dr for flagging.

### LINKED ISSUE #93: Validator does not flag additional fields as invalid
Certain feeds with additional non-GBFS fields are being checked as valid. 
Example: [Citibike](https://gbfs.citibikenyc.com/gbfs/en/station_information.json) includes fields in `station_information.json` like `electric_bike_surcharge_waiver` and `external_id`, but the [validation report](https://gbfs-validator.netlify.app/?url=https%3A%2F%2Fgbfs.citibikenyc.com%2Fgbfs%2Fgbfs.json) says no errors. 

I believe this might actually be an issue with the schemas and not the validator itself. I am opening the issue on both repositories to be sure. 

My initial instinct for a solution would be to add an `additionalProperties: false` to the schemas, but would like to make sure before opening a PR.

### PR DESCRIPTION
**Pull request by tdelmas**
Add non-jsonSchema validations

Checked errors and warnings are listed in README.md

Display a condensed summary of errors and warnings grouped by types and paths

Fix #92

Fix #93 by displaying a warning

Exemple: https://deploy-preview-106--gbfs-validator.netlify.app/validator?url=https%3A%2F%2Fdubai.publicbikesystem.net%2Fcustomer%2Fgbfs%2Fv2%2Fgbfs.json

image

### CODE REVIEW

File: gbfs-validator/__test__/gbfs.v3.0-RC.test.js [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+        return json
+      }
+    }
+
+    let mockRequests = new InvalidMockRequests()
+
+    gbfsFeedServer = mockRequests.build()
+
+    await gbfsFeedServer.listen()
```
Comment (richfab) [inline gbfs-validator/__test__/gbfs.v3.0-RC.test.js:226]: Same suggestion as in https://github.com/MobilityData/gbfs-validator/pull/104#discussion_r1234474674. Thank you 

```suggestion
    await gbfsFeedServer.listen(serverOpts)
```

Comment (tdelmas) [reply gbfs-validator/__test__/gbfs.v3.0-RC.test.js]: (it's fixed now)

**Comment by richfab** [comment]
Happy to see the wording `nonSchema` in the code 👍 
Due to limited resources, we will only be able to take a closer look in a few weeks. 
Please let us know if we are blocking you in any way.
Thank you for your patience 😌

**Comment by richfab** [comment]
Hi @PierrickP and @tdelmas,
Thank you very much for this very valuable contribution 🙇 
I wonder if we could make the interface more consistent for users when there are both errors and warnings.
What do you think about the UI suggestion below?

cc @josee-sabourin @isabelle-dr 

State | Before | After
-- | -- | --
Summary collapsed | image | image
Summary expanded | image | image

**Comment by hbruch** [comment]
Great extension. I just found this PR searching for a check on `duplicate_vehicle_type_id`, which I encounter currently with [this feed](https://gbfs.nextbike.net/maps/gbfs/v2/nextbike_fg/en/vehicle_types.json), and which could be a nice addition to the nonSchema tests.

**Comment by PierrickP** [comment]
@richfab @isabelle-dr i regrouped errors together and update the design

@hbruch I've added a new rule (`duplicate_vehicle_type_id`) to check duplicated `vehicle_type_id` on `vehicle_types.json`

File: gbfs-validator/nonSchemaValidation/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+If a pricing plan id is used (like in `default_pricing_plan_id` or `pricing_plan_ids` of the `vehicle_types.json` file ), but is not present in the `system_pricing_plans.json` file, the validator will return an error.
+
+# default_reserve_time
+
+If the `default_reserve_time` is not present in the `vehicle_types.json` file for a vehicle type that use a pricing declaring `reservation_price_per_min` or `reservation_price_flat_rate`, the validator will return an error.
+
+# num_vehicles_available_incorrect
+
+If `num_vehicles_available_incorrect` on a station is not equal to the sum of `vehicle_types_available.count` for each vehicle type, the validator will return an error.
```
Comment (richfab) [inline gbfs-validator/nonSchemaValidation/README.md:52]: Small typo here. Thanks!

```suggestion
If `num_vehicles_available` on a station is not equal to the sum of `vehicle_types_available.count` for each vehicle type, the validator will return an error.
```

Comment (tdelmas) [reply gbfs-validator/nonSchemaValidation/README.md]: Fixed, thanks!

File: gbfs-validator/nonSchemaValidation/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+If the `vehicle_status.json` contains duplicate `vehicle_id`, the validator will return an error.
+
+# invalid_pricing_plan_id
+
+If a referenced `pricing_plan_id` is not present in the `system_pricing_plans.json` file, the validator will return an error.
+
+# low_cost
+
+If the computed cost seams low regarding the `form_factor` and `propulsion_type`, the validator will return a warning.
```
Comment (richfab) [inline gbfs-validator/nonSchemaValidation/README.md:132]: May we remove `propulsion_type` as it's not used to compute the cost?

```suggestion
If the computed cost seams low regarding the `form_factor`, the validator will return a warning.
```

File: gbfs-validator/nonSchemaValidation/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+# high_cost
+
+If the computed cost seams high regarding the `form_factor` and `propulsion_type`, the validator will return a warning.
```
Comment (richfab) [inline gbfs-validator/nonSchemaValidation/README.md:136]: May we remove `propulsion_type` as it's not used to compute the cost?

```suggestion
If the computed cost seams high regarding the `form_factor`, the validator will return a warning.
```

File: gbfs-validator/__test__/gbfs.v3.0-RC.test.js [hunk 5]
```diff
@@ -151,3 +173,89 @@ describe('exaustive feed', () => {
...
+  let gbfsFeedServer
+
+  beforeAll(async () => {
+    const { MockRequests } = require('./fixtures/v3.0-RC/exaustive')
+
+    class InvalidMockRequests extends MockRequests {
+      system_pricing_plans(...args) {
+        const json = super.system_pricing_plans(...args)
+
+        // This plan with require a default_reserve_time on vehicle_types using it.
```
Comment (richfab) [inline gbfs-validator/__test__/gbfs.v3.0-RC.test.js:187]: Small typo

```suggestion
        // This plan will require a default_reserve_time on vehicle_types using it.
```

**Review by richfab — CHANGES_REQUESTED** [review]
Very impressive work @tdelmas !

Thank you very much for this extremely valuable addition to the validator. I am sure many producers will appreciate this.

Your comments on my suggestions are welcome.

cc @PierrickP @davidgamez

File: gbfs-validator/nonSchemaValidation/README.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+If the `default_pricing_plan_id` is not present in the `vehicle_types.json` file when pricing plans are present, the validator will return an error.
+
+# missing_station_information
+
+If a station is declared `station_status.json` but not in `station_information.json`, the validator will return an error.
+
+# missing_station_status
+
+If a station is declared `station_information.json` but not in `station_status.json`, the validator will return an error.
+
```
Comment (richfab) [inline gbfs-validator/nonSchemaValidation/README.md:41]: May we add the key `duplicate_station_id` used in `nonSchemaValidation/check_stations.js` but missing from `README.md`?

```suggestion

# duplicate_station_id
If `station_information.json` or `station_status.json` contain duplicate `station_id`, the validator will return an error.

```

**Comment by tdelmas** [comment]
Thank you all for your detailed review.

I've pushed a commit that apply most suggestions of your reviews, and marked them as "resolved" for readability.

Comments not marked as resolved are not fixed, but as they require more tests, I strongly feel they are more suited in a follow-up pull request.

**Comment by richfab** [comment]
Thank you @tdelmas for taking most suggestions into consideration. This is amazing!

I will review your changes as soon as possible.

File: gbfs-validator/__test__/fixtures/conditionnal_no_vehicle_type_id.js [hunk 7]
```diff
@@ -1,5 +1,7 @@
 const fastify = require('fastify')
 
+const last_updated_fresh = Math.floor(Date.now() / 1000) - 30
```
Comment (testower) [inline gbfs-validator/__test__/fixtures/conditionnal_no_vehicle_type_id.js:3]: As a general rule of thumb, testing based on an actual clock could lead to unexpected results and false negatives / positives. Using a mocked clock is preferable.

File: gbfs-validator/__test__/fixtures/plan_id.js [hunk 8]
```diff
@@ -1,9 +1,11 @@
 const fastify = require('fastify')
 
+const last_updated_fresh = Math.floor(Date.now() / 1000) - 30
+
 function build(opts = {}) {
   const app = fastify(opts)
 
-  app.get('/gbfs.json', async function(request, reply) {
+  app.get('/gbfs.json', async function (request, reply) {
```
Comment (testower) [inline gbfs-validator/__test__/fixtures/plan_id.js:8]: very nit-picky, but why add a whitespace here?

File: gbfs-validator/__test__/gbfs.v3.0-RC.test.js [hunk 9]
```diff
@@ -137,8 +155,12 @@ describe('exaustive feed', () => {
 
     expect.assertions(2)
 
-    return gbfs.validation().then(result => {
-      expect(get_errors(result)).toEqual([])
+    return gbfs.validation().then((result) => {
+      expect(get_errors(result)).toEqual({
+        errors: [],
+        nonSchemaErrors: [],
+        warnings: []
```
Comment (testower) [inline gbfs-validator/__test__/gbfs.v3.0-RC.test.js:162]: The taxonomy here feels odd. There are errors and there are warnings. But there are also a specific kind of error, which is not from the schema. Where do warnings come from?

My hope is that validation results become standardised for portability reasons, so I hope we can spend the extra effort to consider a rational api.

File: gbfs-validator/nonSchemaValidation/check_vehicle_types.js
Comment (testower) [inline gbfs-validator/nonSchemaValidation/check_vehicle_types.js:1]: Running the risk of repeating myself, again, I find the assumptions made here to produce warnings are highly speculative and I don't think they belong in a validator. They are business rules that depend on operational and geographic circumstances. This just adds unnecessary complexity.

**Review by testower — CHANGES_REQUESTED** [review]
I've left some comments but would like to highlight some issues:

None of the new rules are implemented using partials / schema patching, although after skimming through them, I think a lot of them could well have been (?) - is this a strategic decision? What about the rules that are already implemented that way, are they going to be left like this or ported to the imperative style? The benefit of using the "old" convention is that the output automatically conforms to the regular schema validation error outputs, whereas in the imperative style, they have to be constructed manually. Especially since this project has no type-checking, this is prone to errors.

There is a breaking change in the response format of the validator api - which prompts me to raise the question of standardising this format for portability purposes. If validators respond the same way, frontends and validators can be used interchangeably, and validation reports can be consumed later by any "conformant" UI.

Last, the most serious issue for me, this PR introduces a number of warnings based on assumptions about a number of things that I don't think a generic / canonical GBFS validator should be concerned with. Business rules that depend on operational context should be left out of the validator.

If this validator was a library, one could imagine a feature that allowed library users to provide their own additional rulesets specific to their operational context - but as it stands now, this is just speculative.

Don't get me wrong, I applaud the good intentions behind this, but I don't think it belongs in the validator.

**Comment by tdelmas** [comment]
@testower thank you for your review.

I'll try to answer your concerns.

> None of the new rules are implemented using partials / schema patching, although after skimming through them, I think a lot of them could well have been (?) - is this a strategic decision?

Yes. Logic errors are much more difficult to write using schemas, and even where they are, it leads to some problems and bugs. For example, if two rules needs to patch the same part of a schema, the results will be unpredictable. Or if the original schema changes. Also, most of them are cross-files, which least to reading a file A, then creating the schema for B and vice versa. 

> What about the rules that are already implemented that way, are they going to be left like this or ported to the imperative style?

I feel like some should, in the future.


>The benefit of using the "old" convention is that the output automatically conforms to the regular schema validation error outputs, whereas in the imperative style, they have to be constructed manually. Especially since this project has no type-checking, this is prone to errors.

I agree, this could be improved in a future PR.

> There is a breaking change in the response format of the validator api - which prompts me to raise the question of standardizing this format for portability purposes. If validators respond the same way, frontends and validators can be used interchangeably, and validation reports can be consumed later by any "conformant" UI.

That's a good goal.

> Last, the most serious issue for me, this PR introduces a number of warnings based on assumptions about a number of things that I don't think a generic / canonical GBFS validator should be concerned with. Business rules that depend on operational context should be left out of the validator.

Warning are just warnings: here to catch the attention, because it caught something that may be an error. I insist on the **may**.

I strongly feel they are important because they help to catch subtle mistakes, such as prices in cents instead of base units, or meters vs kilometers, because some feeds had those problems.

> If this validator was a library, one could imagine a feature that allowed library users to provide their own additional rulesets specific to their operational context - but as it stands now, this is just speculative.

A separation could be done, it's an interesting idea for a future PR

File: gbfs-validator/nonSchemaValidation/check_pricing_plans.js
[the remark this answers is not in the document]
Comment (tdelmas) [reply gbfs-validator/nonSchemaValidation/check_pricing_plans.js:1]: They only are done for `USD` and `EUR` https://github.com/tdelmas/gbfs-validator/blob/additional-errors/gbfs-validator/nonSchemaValidation/check_pricing_plans.js#L95

**Comment by testower** [comment]
Let's take a step back to look at first principles. The GBFS validator is, according to the README

> A [General Bikeshare Feed Specification](https://github.com/MobilityData/gbfs) dataset validator

furthermore 

> The Canonical GBFS Validator is a tool to check the conformity of a GBFS feed against the [official specification](https://github.com/MobilityData/gbfs/blob/master/gbfs.md).

The word "canonical" is instructive if taken to mean: "according to the canon", the canon being the "the standard, rule or primary source that is accepted as authoritative"[1]. The "canon" in this context is obviously the GBFS specification.

I will interpret this to mean that the validator is intended to be isomorphic[2] to the specification it validates against. Any departure from that principle should be seriously discussed.

Let me elaborate on why I think it's a bad idea to introduce validation rules that break this isomorphism. My argument is that any validation rule that doesn't have a corollary in the specification, is a *business rule* only relevant for a specific *operational context*.

So while it *may* be useful to get a warning if a car is reported to have room for more than 5 passengers, or a moped has a max range of more than 100km, or a car rental costs more than 50 USD/EUR for an hour (?) - just to name a few examples - these are useful in a specific *operational context* (perhaps yours?), but it's hard to argue that they are universal truisms about the domain over which the GBFS specification is meant to govern. If they were, they would have been encoded in the specification already.

Now I'm not saying that such rules don't have merit, quite the contrary. But I'm saying that they belong in the specific systems that operate in the specific contexts that give rise to their usefulness. Anywhere else, they create clutter and noise.

For the sake of argument, I shall entertain the idea that these rules *should* be in the validator. Now, what process or rationale should determine what these rules and their threshold values should be? Is it consensus-based, or democratic? The greatest common denominator of all producers? Who shall bear the burden of maintaining these over time?

As an exercise, imagine that ever more currencies are added to the "typical cost" structure. How is it possible to keep this structure updated and relevant as prices change and currencies inflate and deflate.

I believe that the best candidates for taking that responsibility are those who operate in the context where those rules, thresholds and values are useful and relevant. Offloading that to the canonical validator is a recipe for future decay and ultimately complete irrelevance of the validator as a general-purpose tool to validate against the *specification*.

If I were to pretend to be positive about accepting these new rules, I would still expect the following preconditions to be met:

* The codebase clearly separates them from other schema or non-schema *specification* rules
* The UI clearly separates the output in the same way. Marking them as warnings is not sufficient. They must be clearly labelled as not being part of the specification validation, and that they should be interpreted as advisory
* Each specific rule should be opt-in as opposed to all of them opt-out (or as it is now, non-optional)
* Each specific rule must be documented both in code and in the UI, the purpose must be clear
* The threshold values should be configurable

Even with these preconditions met, I'm not enthusiastic about them, since they still introduce a maintenance burden. I.e. someone must take responsibility for keeping them up-to-date and relevant, and someone must make sure that the total set of such non-specification rules is kept at a reasonable level. I'm particularly skeptical about the "typical cost" check. For one because it equates USD with EUR which could quickly change - but more importantly because it opens the possibility to add checks for any number of currencies, which, needless to say, would very quickly grow to unmanageable scales. Barring additional currencies from being added to avoid this problem, just amounts to discrimination, so I just don't see a way to accept it in the first place.

Related to the point I made about the output from the validator being standardized, I'm not sure how this fits in. The taxonomy is very unclear and ambiguous about what constitutes a validation error based on the *specification* and what is a result of an opinionated "sanity check".

All in all, I think this marks a very significant shift in the goal and purpose of the validator that we should not take on lightly. I ask the community to think very carefully about accepting these kinds of changes.

[1] https://en.wikipedia.org/wiki/Canonical
[2] Isomorphic in the set-theoretic sense: The set of validation rules should be identical to the set of rules in the specification.

**Comment by tdelmas** [comment]
> So while it may be useful to get a warning if a car is reported to have room for more than 5 passengers, or a moped has a max range of more than 100km, or a car rental costs more than 50 USD/EUR for an hour (?) - just to name a few examples - these are useful in a specific operational context (perhaps yours?)

As I said, those rules are not here to detect "high" or "low" prices, but cases when the producer made a conversion error (between meters vs kilometers, $ vs cents, etc.). But yes, it's not a possible to detect that perfectly. 

@testower Is there a list of rules we can remove from this PR, so it could be acceptable for you?

**Comment by JohanEntur** [comment]
@tdelmas The issue is that there is a **severe break in principle between a format-driven validation, and inclusion of _subjective_ detection of possible user errors**. It's not about what is acceptable for @testower personally, it's about having a whitelabel validator that works equally well for any user. That is the _main point_ here.

Validation rules (the warnings), like these, do have a role to play but they would work better as a **separate validator**, perhaps as a template that each user can configure for local use. That way, each user can define their own limitations in weight, size, doors, and prices according to their own needs - but this still has nothing to do with the GBFS standard.

Our intention is not to be troublesome but to ensure that certain principles are maintained.

**Comment by testower** [comment]
> @testower Is there a list of rules we can remove from this PR, so it could be acceptable for you?

Yes, all the warnings.

**Comment by richfab** [comment]
Hello everyone,

Thank you all for sharing your views on the validator with such enthusiasm and quality arguments. It’s very exciting to see that this tool is getting the attention it deserves from the community.

Thank you @testower for reminding the purpose of the canonical GBFS validator, which is to validate the conformity of a GBFS feed with the specification. MobilityData will make sure to keep this at the forefront of our decisions.

To ensure isomorphism between the specification and the validator, MobilityData recommends continuing to include in the validator only the rules that are in the specification.

The data sanity checks proposed by @tdelmas is a very good idea and could have a place in a separate validator configurable by users, as suggested by @JohanEntur.

Here is MobilityData’s proposal:
1. The validator should return an error for these rules defined as mandatory in the specification:

Rule | Proposal | [GBFS spec](https://github.com/MobilityData/gbfs/blob/master/gbfs.md) extract
-- | -- | --
`missing_translation` | Keep in this PR | "Translations MUST be provided for each supported language for all translatable fields of type Array<Localized String>"
`unknown_language` | Keep in this PR | "Must match one of the values specified by the languages field in system_information.json.”
`missing_default_pricing_plan_id` | Keep in this PR | "REQUIRED if system_pricing_plans.json is defined.”
`missing_station_information` | Keep in this PR | "Any station that is represented in station_status.json MUST have a corresponding entry in station_information.json.”
`missing_station_status` | Keep in this PR | "Any station that is represented in station_information.json MUST have a corresponding entry in station_status.json.”
`duplicate_station_id` | Keep in this PR | "Array that contains one object per station as defined below.”
`default_reserve_time` | Keep in this PR | "REQUIRED if reservation_price_per_min or reservation_price_flat_rate are defined.”
`unclosed_polygon` | Keep in this PR | "The first and last positions are equivalent, and they MUST contain identical values (https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6)”
`duplicate_bike_id` | Keep in this PR | "Array that contains one object per vehicle”
`duplicate_vehicle_id` | Keep in this PR | "Array that contains one object per vehicle”
`invalid_vehicle_type_id` | Keep in this PR | "The vehicle_type_id of this vehicle, as described in vehicle_types.json.”
`duplicate_vehicle_type_id` | Keep in this PR | "Array that contains one object per vehicle type”
`unknown_plan_id`, `missing_system_pricing_plans` and `invalid_pricing_plan_id` | Combine the 3 alerts as they serve the same purpose | "A plan_id, as defined in system_pricing_plans.json”

2. The validator should return a warning for these recommendations defined in the specification:

Rule | Proposal | [GBFS spec](https://github.com/MobilityData/gbfs/blob/master/gbfs.md) extract
-- | -- | --
`last_updated_outdated` | Return a warning instead of an error | "in no case should it be more than 5 minutes out-of-date”
`num_vehicles_available_incorrect` | Return a warning instead of an error | "The total number of vehicles from each of these objects SHOULD add up to match the value specified in the num_vehicles_available field.”
`num_docks_available_incorrect` | Return a warning instead of an error | "The total number of docks from each of these objects SHOULD add up to match the value specified in the num_docks_available field.”
`additional_properties` | Only return a warning when the field name does not start with underscore ( _ ) | "Field names of extensions SHOULD be prefixed with an underscore ( _ ) character”

3. The validator should not return any error or warning for something that is not currently part of the specification: `last_updated_future`, `duplicate_translation`, `duplicate_vehicle_accessory`, `multiple_door_counts`, `num_docks_available_high`, `num_vehicles_available_high`, `station_capacity_too_low`, `unexpected_rider_capacity`, `unexpected_cargo_volume_capacity`, `unexpected_cargo_load_capacity`, `unexpected_propulsion_type`, `unexpectedly_low_range_meters`, `unexpectedly_high_range_meters`, `unexpected_vehicle_accessory`, `unexpectedly_low_wheel_count`, `unexpectedly_high_wheel_count`, `low_cost`, `high_cost`.
We suggest creating a separate issue to keep a list of potential errors/warnings that would be outside of the specification where others can add so that we have a good basis when/if we start to add data sanity checks to the validator.

Please let us know what you all think. 
Thank you!

**Comment by richfab** [comment]
Hi @tdelmas, please let me know if you have any questions or concerns about the proposal. Thank you!

**Comment by tdelmas** [comment]
Thank you for the feedbacks and your ping. 

- I don't have any objections to those changes, I understand the rationale behind it.
- Unfortunately, I don't have the time to work on it at the moment. If any of you want to take over this PR it will be appreciated

**Comment by richfab** [comment]
Thank you @tdelmas and @testower. Since consensus was reached, I am planning to work on removing the code related to data sanity rules from this PR next week. The code will not be lost as it will remain in the commits.

The discussion about data sanity rules was moved to https://github.com/MobilityData/gbfs-validator/issues/138.

Thank you!

**Comment by richfab** [comment]
> if these changes are made I think we can move on to the more technical aspects of the review

@testower @tdelmas A technical review would be very welcome in https://github.com/MobilityData/gbfs-validator/pull/141, if you have the bandwidth.

Thank you in advance for making the GBFS validator even more useful to the community 🙌

**Pull request closed without merging**